### PR TITLE
Pump action and bolt action guns now need to be cycled after firing again

### PR DESCRIPTION
**Pull request by SpeltIncorrectyl**
## About the PR
Fixes issue #19250, where certain guns that should need to be cycled after firing do not need to be. This has occurred as a result of #17219, which removed the Cycled and AutoCycle from BallisticAmmoProviderComponent.cs, and the accompanying code in the relevant systems. This pr adds them back, and sets the correct values in the yaml.

## Why / Balance
These guns are intended to be cycled, and a bug prevented that from being the case.

## Technical details
* Added Cycled and AutoCycle fields to the BallisticsAmmoProviderComponent.cs.
* Altered OnBallisticTakeAmmo method from SharedGunSystem.Ballistic.cs to return early if the BallisticsAmmoProviderComponent is not cycled. This makes it so you cannot fire the gun if it is not cycled.
* Altered OnBallisticTakeAmmo method from SharedGunSystem.Ballistic.cs to set Cycled to false when the gun is fired if AutoCycle is false. This makes it so that pump action or bolt action guns will not be cycled after firing, and will have to be cycled manually to fire (the whole point of this pr).
* Made it so ammo is not ejected from the gun if a manual cycle has cycled the gun. Otherwise you would fire a shotgun, then cycle it so you can fire again, and (unexpectedly) eject ammo from the gun.
This functionality mimics what existed before #17219. 

### Guns Altered ###
* The China Lake
* Kardashev-Mosin
* Hristov
* Musket
* Flintlock Pistol
* ~~Double-Barrelled Shotgun~~ (reverted, can now fire both shots without needing to be cycled)
* Enforcer
* Kammerer
* Sawn-Off Shotgun
* Handmade Pistol
* Blunderbuss
* Improvised Shotgun

**Changelog**

:cl:
- fix: Certain guns (shotguns, launchers, snipers) now need to be cycled after firing again.

### CODE REVIEW

**Comment by deltanedas** [comment]
does this make Z work to rack a gun again, hovering the gun and using e is not very ergonomic

**Comment by JoeHammad1844** [comment]
dont make the DB pump action, please. it should be an advantage of the DB and is more realistic

File: Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs [hunk 1]
```diff
@@ -48,6 +48,22 @@ public sealed partial class BallisticAmmoProviderComponent : Component
...
+    [AutoNetworkedField]
+    public bool Cycled = true;
+
+    /// <summary>
+    /// Automatically cycles the firearm after firing a round
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("autoCycle")]
+    [AutoNetworkedField]
+    public bool AutoCycle = true;
+
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs:66]: Just make it use chambermagazines if you want the functionality back.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by SpeltIncorrectyl** [comment]
> does this make Z work to rack a gun again, hovering the gun and using e is not very ergonomic

Yes, pressing Z with the gun in hand cycles it.

File: Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs (hunk 1 above)
Comment (SpeltIncorrectyl) [reply Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs:66]: What do I do with the chambermagazines? The ChamberMagazineAmmoProvider component requires a magazine for it to function, and none of these guns have magazines.

Comment (metalgearsloth) [reply Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs:66]: I'm having second thoughts on it so let me think about it a bit and I'll come back to you.

**Comment by JoeHammad1844** [comment]
I think they meant toggling the bolt

**Comment by SpeltIncorrectyl** [comment]
> I think they meant toggling the bolt

You mean opening and closing the bolt separately? That is outside the scope of the pr, and is covered in issue #19284.
The thing you have to do after shooting, which this pr changes, can be done with Z.

**Comment by nopeingeneer** [comment]
I mean, leave enforcer alone, it's basically a better shotgun that should have non pump action magazine

**Comment by deltanedas** [comment]
then its just a funny bulldog

**Comment by nopeingeneer** [comment]
I mean, in SS13 there is combat shotgun from cargo that do much the same shit - use semi-automatic across with pump action just like SPAS-12

**Comment by JoeHammad1844** [comment]
syndies should have better equipment than NT, theres more NT personnel than syndicate agents

**Comment by deltanedas** [comment]
also pro idea make double barrel break action instead of pump action
shoot twice with no delay but to reload you have to break it (see: bolting) then eject and load new shells

File: Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs (hunk 1 above)
Comment (SpeltIncorrectyl) [reply Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs:66]: Have you decided anything yet?

Comment (metalgearsloth) [reply Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs:66]: Make cycled nullable instead and update it, then add a method for IsCycled that checks if it's null or true.

Comment (SpeltIncorrectyl) [reply Content.Shared/Weapons/Ranged/Components/BallisticAmmoProviderComponent.cs]: Like this? https://github.com/space-wizards/space-station-14/pull/19287/commits/bb6266dd442e81c0a8918aae5aa5d8e671d75312

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 2]
```diff
@@ -208,6 +210,10 @@ protected int GetBallisticShots(BallisticAmmoProviderComponent component)
 
     private void OnBallisticTakeAmmo(EntityUid uid, BallisticAmmoProviderComponent component, TakeAmmoEvent args)
     {
+        //give up loading if the gun isn't cycled
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:213]:
```suggestion
        // give up loading if the gun isn't cycled
```

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 3]
```diff
@@ -228,6 +234,10 @@ private void OnBallisticTakeAmmo(EntityUid uid, BallisticAmmoProviderComponent c
             }
         }
 
+        //un-cycle the firearm
+        if (!component.AutoCycle)
+            component.Cycled = false;
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:239]:
```suggestion
        component.Cycled = component.AutoCycle;
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
1. Does ItemStatus indicate whether something is cycled.
2. Is there a popup

Doesn't look like there's either but I didn't checkout the pr.

**Comment by Naive817** [comment]
so sad this isn't done..... pump actions are still broken

**Comment by Vaaankas** [comment]
this still needs to be done

**Pull request closed without merging**